src/laws/Law.jl

```diff
@@ -143,73 +277,234 @@ Law{Array{Float64, 0}}(;
 # TODO: create a meaningful example of Law with glacier-shaped matrix
 ```
 """
-struct Law{CACHE_TYPE, F, INIT, FREQ, MAX, MIN, NAME} <: AbstractLaw
+struct Law{
+    CACHE_TYPE,
+    F, FVJPINP, FVJPθ,
+    INIT,
+    FREQ,
+    PFVJP,
+    VJPType,
+} <: AbstractLaw
     f::F
+    f_VJP_input::FVJPINP
+    f_VJP_θ::FVJPθ
     init_cache::INIT
     callback_freq::FREQ
-    max_value::MAX
-    min_value::MIN
-    name::NAME
-
-    # Single, unambiguous inner constructor
-    function Law{CACHE_TYPE}(f, init_cache, callback_freq, max_value, min_value, name) where {CACHE_TYPE}
+    p_VJP::PFVJP
+    VJPType::VJPType
+    max_value::Float64
+    min_value::Float64
+    name::Symbol
+
+    function Law{CACHE_TYPE}(
+        f, f_VJP_input, f_VJP_θ, init_cache, callback_freq, p_VJP, vjpType,
+        max_value, min_value, name
+    ) where {CACHE_TYPE}
         new{
             CACHE_TYPE,
             typeof(f),
+            typeof(f_VJP_input),
+            typeof(f_VJP_θ),
             typeof(init_cache),
             typeof(callback_freq),
-            typeof(max_value),
-            typeof(min_value),
-            typeof(name),
+            typeof(p_VJP),
+            typeof(vjpType),
         }(
             f,
+            f_VJP_input,
+            f_VJP_θ,
             init_cache,
             callback_freq,
+            p_VJP,
+            vjpType,
             max_value,
             min_value,
             name,
         )
     end
 end
 
-# --- OUTER CONSTRUCTORS ---
 
-# 1. Outer constructor with default values for max_value, min_value, name
-Law{CACHE_TYPE}(f, init_cache, callback_freq; max_value = NaN, min_value = NaN, name = :unknown) where {CACHE_TYPE} =
-    Law{CACHE_TYPE}(f, init_cache, callback_freq, max_value, min_value, name)
-
-# 2. Law with inputs (tuple or named tuple), positional
-Law{T}(inputs::Union{Tuple, NamedTuple}, f, init_cache, callback_freq, max_value, min_value, name=:unknown) where {T} =
-    Law{T}(GenInputsAndApply(inputs, f), init_cache, callback_freq, max_value, min_value, name)
-
-# 3. Law with inputs (tuple or named tuple), minimal positional
-Law{T}(inputs::Union{Tuple, NamedTuple}, f, init_cache, callback_freq) where {T} =
-    Law{T}(GenInputsAndApply(inputs, f), init_cache, callback_freq, NaN, NaN, :unknown)
-
-# # 4. Law without inputs, minimal positional
-# Law{T}(f, init_cache, callback_freq) where {T} =
-#     Law{T}(f, init_cache, callback_freq, NaN, NaN, :unknown)
-
-# 5. Law with keyword arguments (with or without inputs)
-Law{T}(; f!, init_cache, callback_freq=nothing, inputs=nothing, max_value=NaN, min_value=NaN, name=:unknown) where {T} =
-    inputs === nothing ?
-        Law{T}(f!, init_cache, callback_freq, max_value, min_value, name) :
-        Law{T}(GenInputsAndApply(inputs, f!), init_cache, callback_freq, max_value, min_value, name)
-
-# --- END OF CONSTRUCTORS ---
+### Declaration of the law with semantic inputs (cf AbstractInput) ###
```

Comment:
All these constructors are getting extremely complex and hard to track. I think we should be very careful and document the most relevant things.